### PR TITLE
Support for nullable types V2

### PR DESCRIPTION
**Pull request by elythyr**
I started with what you did on #55 
I also added the handling of `iterable`, `callable` and `resource` types.

I couldn't update your PR so I made this one, I don't know if you can "merge" them or if we should close one.
Let me know how you wan to proceed.

### CODE REVIEW

File: tests/Integration/Bridge/TolerantParser/Reflection/ReflectionParameterTest.php [hunk 1]
```diff
@@ -138,6 +138,46 @@ function ($method) {
                 );
             },
         ];
+
+        yield 'It reflect iterable type properly' => [
```
Comment (dantleech) [inline tests/Integration/Bridge/TolerantParser/Reflection/ReflectionParameterTest.php]: s/reflect/reflects (not important)

File: tests/Unit/Core/TypeTest.php [hunk 2]
```diff
@@ -83,6 +83,42 @@ public function provideToString()
...
+            '?string',
+        ];
+
+        yield 'Nullable class' => [
+            Type::fromString('?Foobar'),
+            '?Foobar',
+            '?object',
+        ];
+
+        yield 'asd' => [
```
Comment (dantleech) [inline tests/Unit/Core/TypeTest.php]: asd? :)

File: lib/Bridge/TolerantParser/Reflection/TypeResolver/DeclaredMemberTypeResolver.php [hunk 3]
```diff
@@ -2,14 +2,29 @@
...
+use Microsoft\PhpParser\Node\QualifiedName;
 use Microsoft\PhpParser\Token;
 use Phpactor\WorseReflection\Core\Type;
 use Phpactor\WorseReflection\Core\ClassName;
 use Microsoft\PhpParser\Node;
 
 class DeclaredMemberTypeResolver
 {
-    public function resolve(Node $tolerantNode, $tolerantType = null, ClassName $className = null): Type
+    private const WRONG_QUALIFIED_NAME = [
```
Comment (dantleech) [inline lib/Bridge/TolerantParser/Reflection/TypeResolver/DeclaredMemberTypeResolver.php]: should be plural, maybe rename to `RESERVED_NAMES`?

**Comment by dantleech** [comment]
Looks good! Added some naming-related comments, and we need to apply the `./vendor/bin/php-cs-fixer fix` I think.

File: tests/Unit/Core/TypeTest.php (hunk 2 above)
Comment (elythyr) [reply tests/Unit/Core/TypeTest.php]: Don't know it's from yourself :)
Will fix it in `Nullable iterable class`

File: tests/Integration/Bridge/TolerantParser/Reflection/ReflectionParameterTest.php (hunk 1 above)
Comment (elythyr) [reply tests/Integration/Bridge/TolerantParser/Reflection/ReflectionParameterTest.php]: Done

File: lib/Bridge/TolerantParser/Reflection/TypeResolver/DeclaredMemberTypeResolver.php (hunk 3 above)
Comment (elythyr) [reply lib/Bridge/TolerantParser/Reflection/TypeResolver/DeclaredMemberTypeResolver.php]: Done

**Comment by elythyr** [comment]
I have only PHP7.4 on my computer and I can't apply the fixer with it.
I will try to update the dependencies and if it's not enough I will install another version of PHP

**Comment by dantleech** [comment]
No worries  - I can also run it, although I think `php-cs-fixer` works on 7.4, I can run it here.

**Comment by elythyr** [comment]
> vendor/bin/php-cs-fixer fix                                                                                                                                                                                        
PHP needs to be a minimum version of PHP 5.6.0 and maximum version of PHP 7.3.*. 

After `composer update`:
> vendor/bin/php-cs-fixer fix
PHP Fatal error:  Declaration of Symfony\Component\EventDispatcher\EventDispatcher::dispatch($event) must be compatible with Symfony\Contracts\EventDispatcher\EventDispatcherInterface::dispatch(object $event, ?string $eventName = NULL): object in /media/ewilan/php/phpactor/worse-reflection/vendor/symfony/event-dispatcher/EventDispatcher.php on line 51

After changing the version of `php-cs-fixer` in `composer.json` from `~2.15.0` to `^2.15` it's finally OK.

File: tests/Unit/Core/TypeTest.php (hunk 2 above)
Comment (dantleech) [reply tests/Unit/Core/TypeTest.php]: Wonder what I was thinking ... :+1:

File: lib/Core/Type.php [hunk 4]
```diff
@@ -299,4 +303,69 @@ public function __clone()
...
+        }
+
+        return self::class(ClassName::fromString($type));
+    }
+
+    public function asNullable(): self
+    {
+        $instance = clone $this;
+        ;
+        $instance->nullable = true;
```
Comment (dantleech) [inline lib/Core/Type.php]: one semi colon too many

**Comment by dantleech** [comment]
:boom:

**Pull request merged**